### PR TITLE
feat: Initial commit

### PR DESCRIPTION
**Pull request by bradbeam**
Cleans up the PoC work to add in proxy functionality as part of the generated gRPC definition.
The proxy provides the ability to route an incoming request to N targets defined in the
request metadata.

Signed-off-by: Brad Beam <brad.beam@talos-systems.com>

### CODE REVIEW

**Comment by bradbeam** [comment]
Not the best way to show it, but for review purposes, these are snippets of generated code.


```go
type runnerfn func(*proxyOSClient, interface{}, *sync.WaitGroup, chan proto.Message, chan error)

type proxyOSClient struct {
  Conn     OSClient
  Context  context.Context
  Target   string
  DialOpts []grpc.DialOption
}   
  
func proxyOSRunner(clients []*proxyOSClient, in interface{}, runner runnerfn) ([]proto.Message, error) {
  var (
    errors *go_multierror.Error
    wg     sync.WaitGroup
  )
  respCh := make(chan proto.Message, len(clients))
  errCh := make(chan error, len(clients))
  wg.Add(len(clients))
  for _, client := range clients {
    go runner(client, in, &wg, respCh, errCh)
  } 
  wg.Wait()
  close(respCh)
  close(errCh)
    
  var response []proto.Message
  for resp := range respCh {
    response = append(response, resp)
  } 
  for err := range errCh {
    errors = go_multierror.Append(errors, err)
  } 
  return response, errors.ErrorOrNil()
}     
      
type OSProxy struct {
  Provider tls.CertificateProvider
}     
      
func NewOSProxy(provider tls.CertificateProvider) *OSProxy {
  return &OSProxy{Provider: provider}
}     

func (p *OSProxy) UnaryInterceptor() grpc.UnaryServerInterceptor {
  return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
    pkg := strings.Split(info.FullMethod, "/")[1]
    if pkg != "proto.OS" {
      return handler(ctx, req)
    }
    md, _ := metadata.FromIncomingContext(ctx)
    if _, ok := md["proxyfrom"]; ok {
      return handler(ctx, req)
    } 
    ca, err := p.Provider.GetCA()
    if err != nil {
      return nil, err
    }
    certs, err := p.Provider.GetCertificate(nil)
    if err != nil {
      return nil, err
    }
    tlsConfig, err := tls.New(
      tls.WithClientAuthType(tls.Mutual),
      tls.WithCACertPEM(ca),
      tls.WithKeypair(*certs),
    )
    return p.Proxy(ctx, info.FullMethod, credentials.NewTLS(tlsConfig), req)
  }
}
  
func (p *OSProxy) Proxy(ctx context.Context, method string, creds credentials.TransportCredentials, in interface{}, opts ...grpc.CallOption) (proto.Message, error) {
  var (    
    err      error
    errors   *go_multierror.Error
    msgs     []proto.Message
    ok       bool
    response proto.Message
    targets  []string 
  ) 
  md, _ := metadata.FromIncomingContext(ctx)
  // default to target node specified in config or on cli
  if targets, ok = md["targets"]; !ok {
    targets = md[":authority"]
  }
  proxyMd := metadata.New(make(map[string]string))
  proxyMd.Set("proxyfrom", md[":authority"]...)
  
  clients := []*proxyOSClient{}
  for _, target := range targets {
    c := &proxyOSClient{
      Context: metadata.NewOutgoingContext(context.Background(), proxyMd),
      Target:  target,
    } 
    overrideCreds := creds
    conn, err := grpc.Dial(fmt.Sprintf("%s:%d", c.Target, 50000), grpc.WithTransportCredentials(overrideCreds))
    if err != nil {
      errors = go_multierror.Append(errors, err)
      continue
    }
    c.Conn = NewOSClient(conn)
    clients = append(clients, c)
  switch method { 
  case "/proto.OS/Containers":
    var resp *ContainersReply
    msgs, err = proxyOSRunner(clients, in, proxyContainers)
    for _, msg := range msgs {
      resp.Response = append(resp.Response, msg.(*ContainersReply).Response[0])
    }
    response = resp
  case "/proto.OS/Dmesg":
    var resp *DataReply
    msgs, err = proxyOSRunner(clients, in, proxyDmesg)
    for _, msg := range msgs {
      resp.Response = append(resp.Response, msg.(*DataReply).Response[0])
    }
    response = resp
...
func proxyContainers(client *proxyOSClient, in interface{}, wg *sync.WaitGroup, respCh chan proto.Message, errCh chan error) {
  defer wg.Done()
  resp, err := client.Conn.Containers(client.Context, in.(*ContainersRequest))
  if err != nil {
    errCh <- err
    return
  } 
  resp.Response[0].Metadata = &NodeMetadata{Hostname: client.Target}
  respCh <- resp
} 
    
func proxyDmesg(client *proxyOSClient, in interface{}, wg *sync.WaitGroup, respCh chan proto.Message, errCh chan error) {
  defer wg.Done() 
  resp, err := client.Conn.Dmesg(client.Context, in.(*empty.Empty))
  if err != nil {
    errCh <- err
    return
  } 
  resp.Response[0].Metadata = &NodeMetadata{Hostname: client.Target}
  respCh <- resp
}     

```

File: main.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	// Create a wrapped version of the Descriptors and EnumDescriptors that
+	// point to the file that defines them.
+	g.WrapTypes()
+
+	g.SetPackageNames()
+	g.BuildTypeNameMap()
+
+	g.GenerateAllFiles()
+
+	//brad.Printf("%+v", g.Response)
```
Comment (andrewrynhard) [inline main.go]:
```suggestion
```

File: main.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/protoc-gen-go/generator"
+
+	_ "github.com/golang/protobuf/protoc-gen-go/grpc"
+	_ "github.com/talos-systems/protoc-gen-proxy/pkg/proxy"
+)
+
+func main() {
+
+	//brad := log.New(os.Stderr, "funsies", log.Ldate|log.Ltime|log.Lmicroseconds|log.Llongfile)
```
Comment (andrewrynhard) [inline main.go]:
```suggestion
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+RUN curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | \
+    bash -s -- -b /toolchain/bin v1.20.0
+RUN cd $(mktemp -d) \
+    && go mod init tmp \
+    && go get mvdan.cc/gofumpt/gofumports \
+    && mv /go/bin/gofumports /toolchain/go/bin/gofumports
+RUN curl -sfL https://github.com/uber/prototool/releases/download/v1.8.0/prototool-Linux-x86_64.tar.gz | \
+    tar -xz --strip-components=2 -C /toolchain/bin prototool/bin/prototool
+
+# The build target creates a container that will be used to build Talos source
```
Comment (andrewrynhard) [inline Dockerfile]:
```suggestion
# The build target creates a container that will be used to build the source
```

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+RUN go mod verify
+COPY ./pkg ./pkg
+COPY ./main.go ./main.go
+RUN go list -mod=readonly all >/dev/null
+RUN ! go mod tidy -v 2>&1 | grep .
+
+FROM base AS protoc-gen-proxy-build
+ARG SHA
+ARG TAG
+ARG VERSION_PKG="github.com/talos-systems/talos/pkg/version"
```
Comment (andrewrynhard) [inline Dockerfile:49]: Is this needed? This won't actually do anything in the current state.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	g.GenerateAllFiles()
+
+	//brad.Printf("%+v", g.Response)
+	// Send back the results.
+	data, err = proto.Marshal(g.Response)
+	if err != nil {
+		g.Error(err, "failed to marshal output proto")
+	}
+
+	ioutil.WriteFile("./yolo", data, 0777)
```
Comment (andrewrynhard) [inline main.go]: Not sure what to call it, but something other than `yolo`. Something to help me understand what it is by looking at the name of it.

File: pkg/proxy/proxy.go [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+	args.WriteString("*proxy" + serviceName + "Client, ")
+	args.WriteString("interface{}, ")
+	args.WriteString("*sync.WaitGroup, ")
+	args.WriteString("chan proto.Message, ")
+	args.WriteString("chan error")
+	args.WriteString(")")
+	g.P("type runnerfn func" + args.String())
+}
+
+// yolo struct
```
Comment (andrewrynhard) [inline pkg/proxy/proxy.go]:
```suggestion
```

**Review by andrewrynhard — COMMENTED** [review]
Initial pass to maybe give you something to do while I digest the rest.

File: Dockerfile (hunk 4 above)
Comment (bradbeam) [reply Dockerfile:49]: Copied from talos, but honestly there's no reason we should differ on the versioning. It's not currently included in the source but it's something that should get added. Probably a followup pr.

**Comment by bradbeam** [comment]
Added some comments to the generator methods to try to help give a high level overview of each.

**Review by andrewrynhard — APPROVED** [review]

**Pull request merged**